File: Week 9/X87168_en/permutations_2.cpp
/*
PERMUTATIONS(S)
let L be an empty set (of strings)
    if |S| = 1 then
        append s to L
    else
        for i = 1 to |S| do
            for all t in PERMUTATIONS(S[2],..,S[|S|]) do
                append S[i], t to L
    return L
*/


#include <iostream>
#include <vector>
#include <algorithm>
#include <set>
#include <string>
using namespace std;


set<string> permutations_2(string s) {
    set<string> l;
    if (s.length() == 1) {
        l.insert(s);
    }
    else {
        for (int i = 0; i < s.length(); i++) {
            for (auto t : permutations_2(s.substr(0, i) + s.substr(i + 1, s.length() - i + 1))) {
                l.insert(s[i] + t);
            }
        }
    }
    return l;
}


int main() {
    string s;
    cin >> s;
    set<string> res = permutations_2(s);
    for (auto w : res) {
        cout << w << endl;
    }
    return 0;
}

File: Week 9/X87168_en/permutations_2.py
'''
PERMUTATIONS(S)
let L be an empty set (of strings)
    if |S| = 1 then
        append s to L
    else
        for i = 1 to |S| do
            for all t in PERMUTATIONS(S[2],..,S[|S|]) do
                append S[i], t to L
    let A be an empty list (of strings)
    for all w in L do
        append w to A
    sort A
    return A
'''


import sys


def permutations_2(s):
    l = set()
    if len(s) == 1:
        l.add(s)
    else:
        for i in range(len(s)):
            for t in permutations_2(s[:i] + s[i+1:]):
                l.add(s[i] + t)
    A = list()
    for w in l:
        A.append(w)
    A = sorted(A)
    return A


for w in permutations_2(sys.stdin.read()[:-1]):
    print(w)


    

